gpu_tests/test_hf_compatibility.py

```diff
@@ -0,0 +1,194 @@
+    input_ids = torch.cat([torch.tensor([[0]]), input_ids], dim=-1)
+    input_ids = input_ids
+    return input_ids
+
+
+def get_next_token(logits, tokenizer):
+    pred_next_token = torch.argmax(logits[0, -1], -1)
+    next_token = tokenizer.convert_ids_to_tokens([pred_next_token])
+    next_token = next_token[0].replace("Ġ", "")
+    return next_token
+
+
+def setup_vocab_and_merges(model_path):
+    vocab_file = os.path.join(model_path, "gpt2-vocab.json")
+    merges_file = os.path.join(model_path, "gpt2-merges.txt")
+    tokenizer = GPT2Tokenizer(vocab_file, merges_file)
+    tokenizer.save_pretrained(model_path)
+    return vocab_file, merges_file, tokenizer
+
+
+def load_mp_model_and_run_eval(cfg: MetaseqConfig, **kwargs):
+    vocab_file, merges_file, tokenizer = setup_vocab_and_merges(kwargs["model_path"])
+    orig_dims = []
+
+    prompt_ids = []
+    for prompt in prompts:
+        input_ids = tensorize_input(tokenizer, prompt)
+        # Pad sequence to length 32 to avoid Megatron assertion errors
+        orig_dims.append(input_ids.shape[1])
+        input_ids = F.pad(
+            input=input_ids, pad=(0, 32 - input_ids.shape[1], 0, 0), value=1
+        )
+        prompt_ids.append(input_ids)
+
+    prompt_ids = torch.cat(prompt_ids).cuda()
+
+    task = tasks.setup_task(cfg.task)
+
+    def _build_model(cfg, task):
+        # hardcoded to cpu & fp16
+        model = task.build_model(cfg.model).half().cuda()
+        return fsdp_wrap(model)
+
+    with fsdp_enable_wrap(
+        cfg.distributed_training,
+        use_sharded_state=cfg.distributed_training.use_sharded_state,
+    ):
+        models, _model_args, _task = checkpoint_utils.load_model_ensemble_and_task(
+            utils.split_paths(cfg.common_eval.path),
+            arg_overrides=None,
+            task=task,
+            suffix=cfg.checkpoint.checkpoint_suffix,
+            strict=True,
+            num_shards=cfg.checkpoint.checkpoint_shard_count,
+            build_model_hook=_build_model,
+        )
+        model = models[0]
+
+    model.summon_full_params()
+    model = model.eval()
+    model = model
+
+    with torch.no_grad():
+        logits = model(prompt_ids)[0]
+
+    gathered_logits = [
+        torch.zeros_like(logits)
+        for _ in range(dist_utils.get_model_parallel_world_size())
+    ]
+    torch.distributed.all_gather(
+        gathered_logits, logits, group=dist_utils.get_global_group()
+    )
+    gathered_logits = torch.cat(gathered_logits, dim=2)
+
+    # Unwrap gathered logits into separate components for each prompt, and
+    # trim them to match orig_dims
+    trimmed_logits = [
+        logits[:orig_dim].unsqueeze(0)
+        for logits, orig_dim in zip(gathered_logits, orig_dims)
+    ]
+
+    for index, logits in enumerate(trimmed_logits):
+        torch.save(logits, f"/tmp/test_hf_compatibility_{index}.pt")
+
+
+@unittest.skipIf(not torch.cuda.is_available(), "test requires a GPU")
+@unittest.skipIf(
```

Comment:
given that you have the skip here, let's maybe not carve out the special case in the bash setup?

Reply:
Good point... The problem is that we need pytorch 1.9 to run the convert to singleton script as well, which runs in bash, not in the test... Hence I had to add a separate clause in bash.